src/contracts/Output/AdapterNormalizer.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Contracts\Rest\Output;
+
+use Ibexa\Rest\Output\Generator\Json;
+use Symfony\Component\Serializer\Encoder\EncoderInterface;
+use Symfony\Component\Serializer\Normalizer\ContextAwareNormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareTrait;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+final class AdapterNormalizer implements NormalizerInterface, NormalizerAwareInterface, ContextAwareNormalizerInterface
+{
+    use NormalizerAwareTrait;
+
+    private const string CALLED_CONTEXT = __CLASS__ . '_CALLED';
+
+    public function __construct(
+        private readonly EncoderInterface $encoder,
+        private readonly ValueObjectVisitorResolverInterface $valueObjectVisitorResolver,
+    ) {
+    }
+
+    /**
+     * @param array<string, mixed> $context
+     */
+    public function normalize(mixed $object, ?string $format = null, array $context = []): mixed
+    {
+        $eligibleVisitor = is_object($object)
+            ? $this->valueObjectVisitorResolver->resolveValueObjectVisitor($object)
+            : null;
+
+        if ($eligibleVisitor instanceof ValueObjectVisitor) {
+            return $this->visitValueObject($object, $eligibleVisitor);
+        }
+
+        return $this->normalizer->normalize($object, $format, $context);
+    }
+
+    /**
+     * @param array<string, mixed> $context
+     */
+    public function supportsNormalization(mixed $data, ?string $format = null, array $context = []): bool
+    {
+        if (($context[self::CALLED_CONTEXT] ?? false) === true) {
+            return false;
+        }
+
+        $eligibleVisitor = is_object($data)
+            ? $this->valueObjectVisitorResolver->resolveValueObjectVisitor($data)
+            : null;
+
+        if ($eligibleVisitor instanceof ValueObjectVisitor) {
+            return true;
+        }
+
+        return $this->normalizer->supportsNormalization(
```

Comment:
You can assume this will be a `ContextAwareNormalizerInterface`. And even if it isn't, it shouldn't break in current PHP - passing additional arguments is perfectly acceptable language-wise.

However, just to be on the safe side for the application, we should throw a `LogicException` if it's not `ContextAwareNormalizerInterface`. Non context aware variant will go into an infinite loop.